Damcilva/2.0/docs/developer guide v2 (#5218)
microsoft/CBL-Mariner: toolkit/docs/building/building.md

# Developer Guide

## Is this document for you?

|Task                              | Where to go                                                                   |

> |:---------------------------------|-------------------------------------------------------------------------------|

Reviewer: IMHO, we should drop `Just`

> |Just add or build my own packages | **[CBL-MarinerTutorials](https://github.com/microsoft/CBL-MarinerTutorials)** |

Author: I want to direct people to the tutorial repo as much as possible. This is the wrong document for 90% of people to be reading and I want to give them as many offramps as possible.

If you want to experiment with CBL-Mariner, build custom images, or add packages in a more focused environment, refer to the tutorial in the [CBL-MarinerTutorials](https://github.com/microsoft/CBL-MarinerTutorials) repository.

The CBL-Mariner build system consists of several phases and tools, but at a high-level, it can be viewed simply as 3 distinct build stages:

- **Toolchain:** This stage gets a set of toolchain RPM packages needed for future steps. It either downloads stable toolchain packages from our package server, or if configured, builds a bootstrap toolchain and then an official toolchain. Building is highly scripted and serialized in this stage. The official toolchain produced is used in the subsequent package build stage.

- **Package:** This stage uses the toolchain packages from the toolchain stage to build all remaining required RPM packages.  Packages are built in parallel during this stage.

- **Image:** This stage generates ISO, VHD, VHDX, container, etc. images from the RPM packages built in the package stage. The images are defined by configuration.json files.

> It is possible to invoke any stage at any time, the tooling will automatically invoke earlier stages if needed. Each stage can be built completely from scratch, or in many cases may be seeded from pre-built packages and then partially built.

Author: ```suggestion
It is possible to invoke any stage at any time. The tooling will automatically invoke earlier stages if needed. Each stage can be built locally, but in many cases the stage can be seeded from pre-built packages and then partially built.
```

|Work on images               | [Image Stage](#image-stage)                              |
|Work on the toolchain        | [Toolchain Stage](#toolchain-stage)                      |
|Work on the tools            | [Tools](#tools)                                          |
|Build the world from scratch | [Build From Scratch](#building-everything-from-scratch)  |

### Install Prerequisites

Prepare your system by [installing the necessary prerequisites.](prerequisites.md).

### Clone and Sync to a Stable Commit

Clone the 2.0-stable build of CBL-Mariner as shown here.

```bash
# Get the source code
git clone https://github.com/microsoft/CBL-Mariner.git

# Make sure to fetch the latest tags
git fetch --tags

cd CBL-Mariner/toolkit

# Checkout the desired release branch. The 2.0-stable tag tracks the most recent successful release of the 2.0 branch.
git checkout 2.0-stable
```

**NOTE: All subsequent commands are assumed to be executed from inside the toolkit directory.**

### Stable Tags

>_**IMPORTANT:**_ The `2.0-stable` tag always points to the last known good build of the `2.0` branch. \
\
A similar tag, `1.0-stable`, exists for the `1.0` branch. \
\

> Other branches are also buildable but not guaranteed to be stable.  The `1.0` and `2.0` branches are periodically updated with bug fixes, security vulnerability fixes, or occasional feature enhancements.  As those fixes are integrated into the branch the head, of a branch may be temporarily unstable (ie.  only building using the full bootstrap build from source, rather than hydrating from the package servers). \

Author: ```suggestion
Other branches are also buildable but not guaranteed to be stable.  The `1.0` and `2.0` branches are periodically updated with bug fixes, security vulnerability fixes, or occasional feature enhancements.  As those fixes are integrated into the branch, the head of a branch may be temporarily unstable (i.e., only building using the full bootstrap build from source, rather than hydrating from the package servers). \
```

There are several more package build options.  For example, it's possible to build a single package with all of its prerequisites.  For more details on package building options, see [Packages](#packages).

#### Rebuild All Packages

The following command rebuilds all CBL-Mariner packages.

**A reminder that you can build packages outside the core repo (which is often much faster). Refer to the [CBL-MarinerTutorials](https://github.com/microsoft/CBL-MarinerTutorials) repository for examples.**

```bash
# Build ALL packages on a *stable* branch (i.e. 1.0-stable, 2.0-stable, etc)
# (NOTE: CBL-Mariner compiles natively, an ARM64 build machine is required to create ARM64 packages/images)

sudo make build-packages -j$(nproc) CONFIG_FILE="" REBUILD_TOOLS=y
```

```bash
# Build ALL packages on an *unstable* branch
# (NOTE: CBL-Mariner compiles natively, an ARM64 build machine is required to create ARM64 packages/images)

# QUICKREBUILD=y sets both QUICKREBUILD_PACKAGES=y and QUICKREBUILD_TOOLCHAIN=y

# QUICKREBUILD_PACKAGES=y implies the following flags:
#   DELTA_BUILD   = y
#   USE_CCACHE   ?= y
#   REBUILD_TOOLS = y

> # QUICKREBUILD_TOOLCHAIN=y builds the toolchain as quickly as possible, see the toolchain
> # section for details

Author: ```suggestion
# QUICKREBUILD_TOOLCHAIN=y builds the toolchain as quickly as possible by using publicly available packages from our public package repo.
# See the toolchain section for more details.
```

| hybrid approach   | - Best of both        | - May have hard to resolve issues with dependency cycles </br>- Chance that the older toolchain packages hide dependency issues

### Toolchain Stage

The toolchain builds in two sub-phases. The first phase builds an initial _bootstrap_ toolchain which is then used to build the _official_ toolchain used in package building. In the first phase, the bootstrap toolchain downloads a series of sources from our source server. The second uses SRPMS generated from the same sources to build the official RPMs.

For expediency, if `REBUILD_TOOLCHAIN=n` is set, the toolchain will be populated from upstream packages.

#### Populate an Existing Toolchain

A set of bootstrapped toolchain packages (`gcc`, etc.) are used to build CBL-Mariner packages and images. Rather than build the toolchain, the prebuilt binaries can be downloaded to your local machine. This happens automatically when the `REBUILD_TOOLCHAIN=` parameter is set to `n`, which is the default value unless otherwise specified.

```bash
# Populate Toolchain from pre-existing binaries
# This will only work for stable release tags (https://github.com/microsoft/CBL-Mariner/releases)

> sudo make toolchain REBUILD_TOOLS=y

Author: It is the default. I'll add a comment explaining this.

The `CONFIG_FILE` argument provides the way to declare what image you want to build. To manually build **all** packages, you can clear the configuration with `CONFIG_FILE=` and invoke the package build target `make build-packages CONFIG_FILE="" ...`. To build packages needed for a specific image, you must set the `CONFIG_FILE=` parameter to an image configuration file of your choice; the standard image configuration files are in the [./toolkit/imageconfigs](../../imageconfigs/) folder.

Large parts of the package build stage are parallelized; the `Makefile` components may be parallelized by setting the `-j` flag (`-j$(nproc)` works well). The package build scheduler will automatically schedule one build per core, but this can be overwritten using the `CONCURRENT_PACKAGE_BUILDS=` flag.

There are several more package build options; for example, it's possible to build a single package with all of its prerequisites. For more details on package building options, see [Packages](#packages).

#### Rebuild All Packages

The following command rebuilds all CBL-Mariner packages:

> **A reminder that you can build packages outside the core repo (which is often much faster). Refer to the [CBL-MarinerTutorials](https://github.com/microsoft/CBL-MarinerTutorials) repository for examples.**

Author: Moved it above

|Argument|Use|
|:-|-|
| SRPM_PACK_LIST | Only package these `.spec` files for further processing |
| PACKAGE_BUILD_LIST | Build this package, and any dependencies we know about **if they are not already built**|
| PACKAGE_REBUILD_LIST | **ALWAYS** rebuild this package |

```bash
# Clear any existing packages that the tools might know about
sudo make clean-input-srpms clean-expand-specs

# Build targeted packages
packages="openssh at"
sudo make build-packages -j$(nproc) QUICK_REBUILD=y SRPM_PACK_LIST="$packages"
```

Note that this process will download dependencies from [packages.microsoft.com](packages.microsoft.com) and rebuild just the `.spec` files indicated by the `SRPM_PACK_LIST`

After building a package, you may choose to rebuild it or build additional packages.

```bash
# Clean and rebuild targeted packages. Only rebuild the packages if they are not already built.
sudo make clean-build-packages

> sudo make build-packages -j$(nproc) QUICK_REBUILD=y CONFIG_FILE="" SRPM_PACK_LIST="at openssh"

Author: Edge case that I hope to make go away soon anyways so point will be moot. tldr: CONFIG_FILE="..." tells the tools to also build packages for that image definition. It defaults to core-efi.json, which means a bunch of extra stuff gets built. Because we used `SRPM_PACK_LIST` in the first one there is no way for the tools to build other packages to begin with, so it doesn't matter.

Soon I"m going to swap the default to be CONFIG_FILE="".

# - `SOURCE_URL=<YOUR_SOURCE_SERVER>`
# - DOWNLOAD_SRPMS=y (will download pre-packages sources from SRPM_URL_LIST=...)
# - manually placing the correct sources in each /SPECS/* package folder
#     (SRPM_FILE_SIGNATURE_HANDLING=update must be used if the new sources files do not match the existing hashes)
sudo make toolchain PACKAGE_URL_LIST="" REPO_LIST="" DISABLE_UPSTREAM_REPOS=y REBUILD_TOOLCHAIN=y REBUILD_TOOLS=y
```

```bash
# Complete rebuild of all tool, package, and image files from source necessary to build the default image config.

> # NOTE: Source files must made available via one of:

Reviewer 2: must be*
